geco/mips/facility_location/generic.py

```diff
@@ -91,3 +89,87 @@ def capacitated_facility_location(
         model.addCons(customer_facility_vars[i, j] <= facility_vars[j])
 
     return model
+
+
+def capacitated_warehouse_location(
+    n_customers,
+    n_facilities,
+    transportation_cost,
+    demands,
+    fixed_costs,
+    capacities,
+    name="Capacitated Warehouse Location",
+):
+    """
+    Generate a Capacitated Warehouse Location MIP formulation following [1].
+
+    Parameters
+    ----------
+    n_customers: int
+        The desired number of customers
+    n_facilities: int
+        The desired number of warehouses
+    transportation_cost: numpy array [float]
+        Matrix of transportation costs from customer i to warehouse j [i,j]
+    demands: numpy array [int]
+        Demands of each customer
+    fixed_costs: numpy array [int]
+        Fixed costs of operating each warehouse
+    capacities: numpy array [int]
+        Capacities of each warehouse
+    name: str
+        Name of the model
+
+    Returns
+    -------
+    model: scip.Model
+        A pyscipopt model of the generated instance
+
+    References
+    ----------
+    .. [1] J.E. Beasley, An algorithm for solving large capacitated warehouse location problems,
+        European Journal of Operational Research, Volume 33, Issue 3, 1988,
+        Pages 314-325, ISSN 0377-2217,
+        https://doi.org/10.1016/0377-2217(88)90175-0.
+    """
+    model = scip.Model(name)
+
+    model.setMinimize()
+
+    customer_facility_vars = {}
+    facility_vars = []
+
+    # add customer-facility vars
+    for i, j in itertools.product(range(n_customers), range(n_facilities)):
+        var = model.addVar(
+            lb=0, ub=1, obj=transportation_cost[i, j], name=f"x_{i}_{j}", vtype="B"
+        )
+        customer_facility_vars[i, j] = var
+
+    # add facility vars
+    for j in range(n_facilities):
+        var = model.addVar(lb=0, ub=1, obj=fixed_costs[j], name=f"y_{j}", vtype="B")
+        facility_vars.append(var)
+
+    # add constraints
```

Comment:
It would be nice to have the numbering from the paper before each constraint. E.g. (2), for easier mapping.

Reply:
We should also mention why we skip some constraints.